### PR TITLE
doAction when a deprecated feature is encountered

### PR DESCRIPTION
**Pull request by aaronjorbin**
## Description
Whenever a deprecated feature is encountered, an action should be fired.  This is necessary for [log deprecated notices](https://wordpress.org/plugins/log-deprecated-notices/) to be able to easily log these deprecations or for other plugins to handle and do something when somethign deprecarted is encountered.

## How has this been tested?
Added this js to the page:
`wp.hooks.addAction( 'wp.deprecated', 'yo.dawg', function( feature, options, message ){
alert( 'DEPRECATED!' );
console.log( feature, options, message );
});`

There is also a unit test.  ¯\\_(ツ)_/¯

## Types of changes

This adds a new dependency to the deprecated package. Not sure if this needs to be tested specially with packages or we need to do something special. 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by aaronjorbin** [comment]
Simple code I used to test this:

wp.hooks.addAction( 'wp.deprecated', 'yo.dawg', function( feature, options, message ){
    alert( 'DEPRECATED!' );
    console.log( feature, options, message );
});

File: packages/deprecated/src/index.js
Comment (gziolo) [inline packages/deprecated/src/index.js]: I think this could be shortened to:
```js
export default function deprecated( feature, options = {} ) {
const {
	version,
	alternative,
	plugin,
	link,
	hint,
} = options;
```

File: packages/deprecated/src/index.js
Comment (gziolo) [inline packages/deprecated/src/index.js]: It should have:
```js
/**
 * WordPress dependencies
 */
```
comment above.

File: packages/deprecated/src/index.js
Comment (gziolo) [inline packages/deprecated/src/index.js]: I'm wondering if there is a way to turn `options` into shape to avoid duplication. I saw something like this in the past:
https://github.com/WordPress/gutenberg/blob/master/packages/components/src/autocomplete/index.js#L93-L105

**Comment by gziolo** [comment]
It looks like a good addition which would allow to integrate it with other parts of WordPress. Let's run it through weekly JS chat to make sure everybody is on board with the idea.

**Comment by aduth** [comment]
>Let's run it through weekly JS chat to make sure everybody is on board with the idea.

[Added to agenda](https://docs.google.com/document/d/195qmbcZe8GoE1WwQji5dLH64Took5Lb9Fa-MCkyrJY4/edit).

**Review by adamsilverstein — COMMENTED** [review]
Haven't actually tested code, but overall I love this idea. Great work @aaronjorbin!

File: packages/deprecated/src/index.js
Comment (aaronjorbin) [reply packages/deprecated/src/index.js]: done

File: packages/deprecated/src/index.js
Comment (gziolo) [reply packages/deprecated/src/index.js]: Yes, we didn't come up with any standards around that. It might make sense to follow PHP implementation though.

File: packages/deprecated/src/index.js
Comment (gziolo) [inline packages/deprecated/src/index.js]: In the last line of the comment, there is some spacing issue. I think there are used tabs instead of spaces.

File: packages/deprecated/package.json
Comment (gziolo) [inline packages/deprecated/package.json]: We still need to run `npm install` to update `package-lock.json`.

**Review by gziolo — COMMENTED** [review]
Can we also add documentation in `README` file and a unit test before we merge this one?

**Comment by aduth** [comment]
What's the process by which we should decide when it's appropriate to introduce a new action? Is it simply a "that could be useful" consensus? Is there documentation for this even in how PHP hooks are introduced?

**Comment by aaronjorbin** [comment]
> What's the process by which we should decide when it's appropriate to introduce a new action? Is it simply a "that could be useful" consensus? Is there documentation for this even in how PHP hooks are introduced?

There isn't a formal one for PHP Hooks right now. It's generally up to the committers if they think that the use case is one that can only be accomplished if a new hook is added and if that hook is unlikely in their mind to cause a loss of flexibility in the future.

File: packages/deprecated/src/index.js
Comment (aaronjorbin) [reply packages/deprecated/src/index.js]: fixed. thanks

File: packages/deprecated/package.json
Comment (aaronjorbin) [reply packages/deprecated/package.json]: done

**Comment by aaronjorbin** [comment]
@gziolo Documentation and a unit test have been added.

File: packages/deprecated/README.md
Comment (gziolo) [inline packages/deprecated/README.md]: Can we add an example here?

**Review by gziolo — COMMENTED** [review]
My only remaining concern is the name of the action. In other places when we set up hooks we prefix them with the name of the package. In this case, the function is the default export for the package so I think it should be simply:

```js
doAction( 'deprecated', feature, options, message );
```

Let's agree on that before we merge. It would be nice to include an actual example in README.md file, too. Otherwise, it looks great.

**Comment by youknowriad** [comment]
How do we proceed here, can we update the action name to `deprecated` to match the other actions we have (until we reach a consensus on the conventions). We're trying to release 3.7 today, so might be good to update to ship in the release.

**Comment by gziolo** [comment]
I will take care of it 👍

**Review by gziolo — APPROVED** [review]
I addressed my own comments and adjusted the hook name according to the chat we had earlier this week during Core JS weekly meeting.

**Comment by gziolo** [comment]
Thanks @aaronjorbin for leading this PR 🙇

**Pull request merged**
